tools/classes/AssetsManager.ts

```diff
@@ -0,0 +1,438 @@
+import { createRequire } from "node:module";
+import { extname, join, resolve } from "node:path";
+import { fileURLToPath } from "node:url";
+import { tmpdir } from "node:os";
+import {
+	createReadStream,
+	createWriteStream,
+	readFileSync,
+	writeFileSync,
+} from "node:fs";
+import { pipeline } from "node:stream/promises";
+
+import got from "got";
+import glob from "glob";
+import FormData from "form-data";
+import sharp from "sharp";
+
+import { Metadata } from "./PresenceCompiler";
+
+const require = createRequire(import.meta.url),
+	rootPath = resolve(fileURLToPath(new URL(".", import.meta.url)), "../.."),
+	cdnBase = "https://cdn.rcd.gg";
+
+export default class AssetsManager {
+	cwd: string;
+	presenceFolder: string;
+
+	constructor(
+		public service: string,
+		public options?: {
+			cwd?: string;
+		}
+	) {
+		this.cwd = options?.cwd ?? rootPath;
+		this.presenceFolder = this.getPresenceFolder();
+	}
+
+	get assetBaseUrl() {
+		return `${cdnBase}/PreMiD/${encodeURI(
+			this.presenceFolder.replace(this.cwd, "")
+		).replace("#", "%23")}/assets`;
+	}
+
+	getPresenceFolder() {
+		return glob
+			.sync(`{websites,programs}/**/${this.service}`)[0]
+			.replace(/\\/g, "/");
+	}
+
+	getFileExtension(url: string) {
+		return extname(url);
+	}
+
+	get allTsFiles() {
+		return glob
+			.sync(`{websites,programs}/**/${this.service}/**/*.ts`, {
+				absolute: true,
+			})
+			.filter(file => !file.endsWith(".d.ts"));
+	}
+
+	get metadata(): Metadata {
+		return require(resolve(this.presenceFolder, "metadata.json"));
+	}
+
+	/**
+	 * Gets all assets used in the presence
+	 *
+	 * Includes the logo, thumbnail, and all other assets used in the presence
+	 *
+	 * @note Includes both assets uploaded to the cdn, and assets that are not uploaded to the cdn
+	 *
+	 * @returns A list of all assets used in the presence
+	 */
+	get allAssets(): {
+		logo: string;
+		thumbnail: string;
+		assets: Set<string>;
+	} {
+		const assets = new Set<string>(),
+			{ logo, thumbnail } = this.metadata,
+			files = this.allTsFiles;
+
+		for (const tsfile of files) {
+			const file = readFileSync(tsfile, "utf8");
```

Comment:
Maybe don't block the thread? (Doesn't matter in this case but since everything is async might as well)

Reply:
Can't use async in getters.

Reply:
Getters can still return Promises, right?